### PR TITLE
rewrite as ssb-encryption-format

### PR DESCRIPTION
**Pull request by staltz**
Part of https://github.com/ssbc/ssb-db2/pull/347.

Ready for review/merge but before putting this into use I'd like to get ssb-keyring into this module, so that keys are persisted.

@arj03 There's quite a lot of tests that I deleted because they were just using ssb-db2 APIs to test the outcomes, I found ways of testing that don't depend on using ssb-db2. But I'm open to restoring those deleted tests, either here or moving to ssb-db2 repo.

@arj03 I'm thinking of renaming this package (instead of creating a new one that could be conflicting/confusing with this one) from `ssb-db2-box2` to `ssb-box2` since it *could* be useful also outside of ssb-db2. What do you think?

TODO:

- [x] Fix tests in CI
- [x] Update README
- [x] Use new ssb-keyring
- [ ] Add ability to remove group key (useful at least for ssb-db2 benchmarks)

### CODE REVIEW

**Comment by arj03** [comment]
I'm fine with the name change. And the tests looks good. This seems like a good separation of concerns.

One thing that I don't understand with this refactor is that `onReady` was removed. Instead there is an internal setup function that gets calls by db2, but I don't see anywhere where I can, as a user of this module, make sure that all keys are added before we are "ready". Like how do you handle the case where you start indexing things before all keys are added? I'm not sure ssb-keyring is the (only) answer because we would like to use the meta-feeds derived seed, so you would have to wait for that module to load before you have the seed e.g.

**Comment by staltz** [comment]
Good questions @arj03 

I thought that this case could be handled via ssb-keyring, and then in ssb-box2's `setup(cb)` it would load all of ssb-keyring into memory, and only after that's done we call `cb`. This means that keys are available either: (1) right after you add them, or (2) upon startup via `setup(cb)`. I'm assuming that ssb-keyring will be not too slow, at least not slower (hopefully) than `sbot.db.query` (which ssb-meta-feeds needs for loading the seed). Also, loading the seed from ssb-meta-feeds requires waiting for ssb-db2 initialization to be completed, and this includes the private index, so we could be creating a chicken and egg problem.

It seemed simpler to put all known keys in the ssb-keyring which can act as a cache, and it simplifies the code too, because now without `setReady` this module is responsible for its own keys. I think `setReady` could become a footgun because it's not clear who is responsible for defining that box2 is ready to decrypt. With `setup()` it's clear that box2 is ready once it has loaded all ssb-keyring.

This is of course missing the actual implementation with ssb-keyring, but I'm assuming that as soon as you `add` a key, it'll persist it to ssb-keyring such that next time when the app starts up, you don't have to add that key again.

**Comment by staltz** [comment]
> Like how do you handle the case where you start indexing things before all keys are added?

Oh, and about this one. Let's assume we start with zero box2 keys, then as soon as we add the first key, it'll trigger reindex and will persist that key in keyring. Once the app restarts, it'll wait for keys to be loaded from ssb-keyring before doing any kind of indexing (because ssb-db2 waits for leveldb and private indexes first).

**Comment by arj03** [comment]
Ok we can try that, I'll review again once ssb-keyring is part of this PR then :-)

File: format.js [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+  const groupRecpsCount = validRecps.filter(_isGroup).length
+  if (groupRecpsCount > 1) {
+    // prettier-ignore
+    throw new Error(`private-group spec only supports one group recipient, but you've tried to use ${groupRecpsCount}`)
+  }
+
+  const encryptionKeys = validRecps.reduce((acc, recp) => {
+    if (recp === selfId) return [...acc, ..._keysManager.ownDMKeys()]
+    else if (_isGroup(recp)) return [...acc, _keysManager.groupKey(recp)]
+    else if (_isFeed(recp)) return [...acc, _keysManager.sharedDMKey(recp)]
```
Comment (mixmix) [inline format.js:79]: this sharedDMKey is gonna depend on who the "author feed" is.

Comment (mixmix) [reply format.js:79]: selfId...

This `encrypt` is gonna be per-feed I guess... so is this are you registering this format per-feed?

Comment (mixmix) [reply format.js:79]: Need to get my head around how this fits into the publish pipeline

Comment (mixmix) [reply format.js:79]: Assume it will be something like

```js
ssb.db.createAs(myChessKeys, { content, encryptionFormat: 'box2' }, cb)
```

Comment (staltz) [reply format.js:79]: Yes, but not `createAs`:

```js
ssb.db.create({ keys: myChessKeys, content, encryptionFormat: 'box2' }, cb)
```

Comment (staltz) [reply format.js:79]: The "author feed" is given by `opts.keys.id` and this refers to the keys that are signing/creating the message.

**Comment by staltz** [comment]
This implementation of ssb-box2 is now using a [fork of ssb-keyring](https://gitlab.com/ahau/lib/ssb-keyring/-/tree/staltz-metakeyring), namely branch `staltz-metakeyring`.

There are a couple of surprising things, noteworthy:

## Tests

All the tests for ssb-box2 pass :heavy_check_mark: locally for me, but CI is shouting some errors, I don't know why but I'm gonna get to the bottom of that later.

Tests for ssb-keyring were not touched at all yet, so they're definitely not passing. I built the implementation of ssb-keyring until ssb-box2 tests passed, just to make sure whatever I'm building in ssb-keyring is immediately useful for ssb-box2. 

## Sync init

Initializing keyring is like this:

```js
const Keyring = require('ssb-keyring')

const keyring = Keyring(pathHere, onReady)
keyring.group.add(groupId, info, cb) // already usable, sync
```

Why? Because it's useful synchronously to be able to call such `add` APIs after starting up an sbot *without* having to wait for onReady (which is just about loading existing keys, which is not interesting to us if we're trying to `add` keys), because anyway the .add APIs operate synchronously on the **caches**.

## Own set/get

Own DM keys are now set/get, which means there can only be one DM key. Why? Because when you're *encrypting* a message, we would be in big trouble if you had, say, 17 self DM keys. But in less extreme cases, you might have exactly 16 recps, where one of them is you, but because you have **two** self DM keys, the total "actual" encryption keys would be 17.

## Keypair adding

This is a new concept where you can add `{ public, private }` objects directly from ssb-keys. This will be useful to load all the metafeed-tree keys you own, and being able to say that any given feedId is "mine".

## New API

The API of that branch is:

- `dm`
  - `.get(authorId, feedId)` INTERNAL
  - `.list(feedId)` INTERNAL
- `own`
  - `.set(info, cb)`
  - `.get()` INTERNAL
- `group`
  - `.add(groupId, info, cb)`
  - `.has(groupId)`
  - `.get(groupId)` INTERNAL
  - `.list()` INTERNAL
- `keypair`
  - `.add(authorId, keys, cb)`
  - `.get(authorId)` INTERNAL
  - `.has(authorId)` INTERNAL
  - `.list()` INTERNAL
- `encryptionKeys(authorId, recps)`
- `decryptionKeys(authorId)`
- `close(cb)`

**Comment by arj03** [comment]
I wonder if it would make sense to be able to remove group keys again

**Comment by mixmix** [comment]
yes to removing. But later!

thanks @staltz The only thing I disagree with is that `own.set`. You may want to change your own keys, but still be able to decrypt past ones. I think the API there should be 

```js
own.add(feedId, info, cb)   // NOTE this is associated with a feedId !
own.get(feedId) // gets latest encryption key for if you set recps: [feedId]
own.list(feedId) // get all ownKeys ever set for a feedId, as an Array
```

**Comment by staltz** [comment]
@mixmix but how do you address the problem of more than 16 slots used?

And why should the own DM key be *per feedId*?

> You may want to change your own keys, but still be able to decrypt past ones.

In what cases we would need/want this?

**Comment by mixmix** [comment]
The envelope spec let's you have as many slots as you want. The downside is that the more slots they're are, the more checks you have to do. So if we set a max of 64 slots we would have to check up to 64 times per message. Slow.

So, in the private group spec for ssb, we wrote that the max slots should be 16.

**Comment by mixmix** [comment]
i.e. Any validators of recps needs to reject more than 16 recps, and any implementation of private groups encryption should also reject.

**Comment by mixmix** [comment]
> Why should the self key be associated with a feedId

So that when you read that feedId in the recps you know which self key to try. I think what you're suggesting that internally if you see a feedId and you answer "is this one of my feedIds?" (in my tree of meta feeds) ,then if the answer is "yes" you could just use the one self Key.

The performance enhancement only lands when you have less own keys to try on decryption. If you have one selfKey pet feed, then there is still only one selfKey to try for messages in that feed. The difference is that you could have a feed where you know you are not gonna self send (like in the bat's group where you only use the group id).

I'm this case you could have zero self keys to try. Suuuper fast.

**Comment by staltz** [comment]
> i.e. Any validators of recps needs to reject more than 16 recps, and any implementation of private groups encryption should also reject.
> 
> 

The problem is that you may have 1 recp, which is a feedId you own, but then that 1 recp gets mapped to N encryption keys. N potentially large.

**Comment by staltz** [comment]
> > Why should the self key be associated with a feedId
> 
> So that when you read that feedId in the recps you know which self key to try. I think what you're suggesting that internally if you see a feedId and you answer "is this one of my feedIds?" (in my tree of meta feeds) ,then if the answer is "yes" you could just use the one self Key.
> 
> The performance enhancement only lands when you have less own keys to try on decryption. If you have one selfKey pet feed, then there is still only one selfKey to try for messages in that feed. The difference is that you could have a feed where you know you are not gonna self send (like in the bat's group where you only use the group id).
> 
> I'm this case you could have zero self keys to try. Suuuper fast. 

I went with 1 self DM key per metafeed-tree not for performance reasons, but for YAGNI reasons and to solve the 16 slot limit problem.

**Comment by mixmix** [comment]
> The problem is that you may have 1 recp, which is a feedId you own, but then that 1 recp gets mapped to N encryption keys. N potentially large.

WHAT? this is very surprising, why would 1 recp map to >1 encryption keys. I don't think we should ever do this

**Comment by staltz** [comment]
@mixmix exactly my point: there should only be 1 own DM key. Why do we need to support many own DM keys when in the end we always use 1?

**Comment by mixmix** [comment]
because we may have past DM keys we want to be able to read, though not encrypt to.
e.g. I might end up with one encryption key which is shared by several devices (via a fusion identity)... though that needs some care and thought....

Why would we cycle them? You may not want to, but I want to build this such that I can cycle the use of them

**Comment by staltz** [comment]
@arj03 Take a look at commit https://github.com/ssbc/ssb-db2-box2/pull/7/commits/a6fe73f56a79750027007680fbc0940b0677d0ab which uses ssb-keyring 2.0.0. This should be the final state of this PR (apart from a version bump once https://github.com/ssbc/ssb-private-group-keys/pull/14 is merged).

**Comment by arj03** [comment]
Exciting to see this moving closer to being ready :)

**Comment by staltz** [comment]
@arj03 This PR is finally ready and it's the final domino to fall.

We don't have "remove group key" at the moment, but perhaps we can do that separately, as (1) add that to ssb-keyring, (2) add that API to ssb-box2, (3) use it in ssb-db2 bench.

File: package.json [hunk 2]
```diff
@@ -32,18 +31,21 @@
     "rimraf": "^3.0.2",
     "secret-stack": "6.3.2",
     "ssb-backlinks": "^2.1.1",
+    "ssb-buttwoo": "^0.3.0",
     "ssb-caps": "1.1.0",
     "ssb-db": "^20.4.0",
-    "ssb-db2": "^2.7.0",
+    "ssb-db2": "ssbc/ssb-db2#formats-split",
```
Comment (arj03) [inline package.json:37]: We have to remember to change this once that branch is merged :)

**Pull request merged**